crates/compat/src/http.rs

```diff
@@ -48,3 +132,73 @@ impl Default for Response {
         }
     }
 }
+
+impl Response {
+    pub async fn from_http(
+        response: http::Response<impl AsyncRead + Unpin>,
+    ) -> anyhow::Result<Response> {
+        let (
+            http::response::Parts {
+                status, headers, ..
+            },
+            mut body,
+        ) = response.into_parts();
+        let status_code = status.as_u16();
+        let content_length = headers
+            .get(CONTENT_LENGTH)
+            .and_then(|content_length| content_length.to_str().ok())
+            .and_then(|content_length| content_length.parse().ok());
+        let header = headers
+            .into_iter()
+            .map(|(name, value)| {
+                let name = name.context("invalid header name")?;
+                let value = value
+                    .to_str()
+                    .with_context(|| format!("failed to parse `{name}` header value as string"))?;
+                Ok((name.as_str().into(), vec![value.into()]))
+            })
+            .collect::<anyhow::Result<_>>()
+            .context("failed to process headers")?;
+        let body = if let Some(content_length) = content_length {
+            let mut buf = Vec::with_capacity(content_length);
+            body.take(content_length.try_into().unwrap_or(u64::MAX))
+                .read_to_end(&mut buf)
+                .await
+                .context("failed to read response body")?;
+            buf
+        } else {
+            let mut buf = vec![];
+            body.read_to_end(&mut buf)
+                .await
+                .context("failed to read response body")?;
+            buf
+        };
+        Ok(Response {
+            status_code,
+            header,
+            body,
+        })
+    }
+}
+
+impl TryFrom<Response> for http::Response<Vec<u8>> {
+    type Error = anyhow::Error;
+
+    fn try_from(
+        Response {
+            status_code,
+            header,
+            body,
+        }: Response,
+    ) -> Result<Self, Self::Error> {
+        let res = http::Response::builder().status(status_code);
+        let res = header
+            .into_iter()
+            .filter_map(|(name, mut values)| {
+                let value = values.pop()?;
```

Comment:
Same question here, can we add context?

Reply:
No, because this needs to return an `Option` and I think that empty header values should just be ignored